### ISSUE TITLE
Fix HTML event-loop

### ISSUE THREAD

**Issue by gterzian** [servo/servo#31242]
Our current implementation of the [HTML event-loop](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model) is in need of fixing. For this issue I am referring only to the window event-loop, which requires interacting with the rendering. Worker loops seem ok for now(we're not delivering rafs to workers yet).

--------------

Quick recap on how it is supposed to work:

1. The event-loop runs a task sequentially. 
2. It runs a micro-task checkpoint(so one after each task is run)
3. It  reports  long tasks.
4. It runs the idle callback logic.

In parallel(which means anywhere else in Servo, like the compositor), we must wait for for a [rendering opportunity](https://html.spec.whatwg.org/multipage/#event-loop-processing-model:rendering-opportunity) for that event-loop, and then queue a task that runs the [update the rendering steps](https://html.spec.whatwg.org/multipage/#update-the-rendering) 

The task queued is then appended to a task queue, and eventually handled at 1 above. Note that this tasks runs a lot of steps, but atomically within the same task, and in a specific order. 

For example, step 11 is [update animations and send events](https://drafts.csswg.org/web-animations-1/#update-animations-and-send-events), step 14 is [update animations and send events](https://drafts.csswg.org/web-animations-1/#update-animations-and-send-events), and step 15 runs the resize observer logic. 

Running these "update the rendering" steps in that way allows for a certain level of determinism, for example if one requests an animation frame callback from within the last notification callback of a resize observer, one can be certain that by the time the next resize notification is callback, the animation frame callback will have been invoked first. This [resize observer test](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/tests/wpt/tests/resize-observer/eventloop.html) is an example of relying on that ordering. 

--------------

Quick recap on how it current works:
1. We [select](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/components/script/script_thread.rs#L1490) on various channels, including one representing the various task queues, but also other channels representing messages from the constellation/compositor, the image cache, webgpu, and others. 
2. We handle certain type of messages [first](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/components/script/script_thread.rs#L1519). 
7. We then [check for additional messages](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/components/script/script_thread.rs#L1605)
8. We then [`update_animations_and_send_events`](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/components/script/script_thread.rs#L1627)
9. We then process events collected at 1 and 3 above [sequentially](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/components/script/script_thread.rs#L1631)
10. We then perform a [micro task checkpoint](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/components/script/script_thread.rs#L1681)
11. We then [`maybe_queue_document_completion`](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/components/script/script_thread.rs#L1689)
12. Finally, we [issue any batched reflows](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/components/script/script_thread.rs#L1699)

The [ResizeObserver PR](https://github.com/servo/servo/pull/31108) adds a step before the reflows, that runs the resize observer notification steps. 

--------------

The problem: our current implementation does not follow the ordering of the spec, hence it is a source of timing issues. 

Solution: restructure the event-loop to only run the 4 steps it is supposed to run, and restructure the handling of all other messages as happening via a tasks running the "update the rendering" steps.

**Comment by KiChjang** [comment]
I can't help but feel like we're trying to re-implement an async runtime on Servo, but without using futures. Has there been any discussion around using async Rust and its related crates to implement the event loop?

**Comment by KiChjang** [comment]
My naïve thinking is that we can just make sure any JS pointer used by a task is rooted, although this likely comes at a perf cost. Aside from that, I think it's also true that the GC is free to move objects around, so I'm not sure whether making everything rooted is sufficient. Futures in Rust also requires all the pointers to be `Pin`ned as well, so this is indeed tricky.

**Comment by gterzian** [comment]
The first thing I will do, is re-shuffling code so as to wrap all the "update the rendering" logic inside a single task(as described above) to match the spec. 

Regarding using async, my current point of view is that it's not a good idea, because:

- The script thread already implements a custom "async runtime", one that runs Javascript code asynchronously and sequentially. Async Rust is not about running Javascript, it is about running Rust. We could restructure our "script thread" as an async "script task" using async Rust, but it's not clear to me what problem that would solve. 
- Specifically regarding integration with Spidermonkey and garbage collection: yes I agree it would be very difficult, in part because  most Rust async tasks come with a `Send` bound([example](https://docs.rs/tokio/latest/tokio/task/fn.spawn.html) from tokio tasks), and, with the exception of the [`Trusted`](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/script/dom/bindings/refcounted.rs#L77) stuff, our current rooting stack assumes a single thread. 
- The HTML event-loop runs tasks sequentially, and multiple event-loops will in practice run in different processes. Async Rust is mostly about running tasks on a threadpool(although single-threaded variants exists). The multi-process nature of modern web engines would exclude the possibility of running multiple HTML event-loops onto a single Rust async runtime. Also, although I don't have a link handy, Chromium have had issues with spec compliance and their async task-like concept of "runnables". Implementing an HTML event-loop within a Rust async runtime would require some way of ensuring sequential execution of tasks, i.e. wrapping task queues in a mutex, so the end-result would be equivalent to running the HTML event-loop on a single thread(but in my opinion more complicated).
- Since doing such a restructuring would be a huge task(no pun intended), it does not seem worth it. 

As a side note, I think there is one part were async could be more useful: the public API of the servo lib, so as to enable the embedding layer to be async and select on different async event sources(as opposed to [draining vectors](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/servo/lib.rs#L832) of event slike we currently do), and also to make it easier for embedders to run Servo within an async context. The problem this would solve is the complexity of the main embedder thread, whose [event-loop logic](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/ports/servoshell/app.rs#L60) is hard to follow because it consists of multiple "event-handling points", see all the `get_events` and `handle_events` calls found in [`App::handle_events`](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/ports/servoshell/app.rs#L430).
A single async select loop would be a big gain in terms of simplicity, and also since async is so poplular, many embedders may be running in an async context to begin with(should open a dedicated issue to discuss). 

cc @wusyong

**Comment by KiChjang** [comment]
One of the problems that I keep encountering is the ability to express the idea to pause execution of the current task and run others that are in the queue, until the current task can resume based on certain conditions.
This is easily expressed with `await` in async Rust, but AFAIK this still cannot be expressed in the current event loop implementation. Perhaps what we need is some sort of function or syntax to perform suspension however, we don't necessarily need to import async Rust wholesale.

**Comment by wusyong** [comment]
> As a side note, I think there is one part were async could be more useful: the public API of the servo lib, so as to enable the embedding layer to be async and select on different async event sources(as opposed to [draining vectors](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/components/servo/lib.rs#L832) of events like we currently do), and also to make it easier for embedders to run Servo within an async context. The problem this would solve is the complexity of the main embedder thread, whose [event-loop logic](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/ports/servoshell/app.rs#L60) is hard to follow because it consists of multiple "event-handling points", see all the `get_events` and `handle_events` calls found in [`App::handle_events`](https://github.com/servo/servo/blob/e76770202c059202889bdb3555e9efb817b6f68e/ports/servoshell/app.rs#L430). A single async select loop would be a big gain in terms of simplicity, and also since async is so poplular, many embedders may be running in an async context to begin with. Note that this does not mean that even `components/servo` should be async, but rather that only its public API could consists of futures and streams(could open a dedicated issue to discuss).

I'm not satisfied with `get_event`/`handle_event` API either, but I think async will make it worse. There will be embedding use cases which prefer single thread event loop and limited resources. I also hope the interface could be a single select loop, but perhaps servo lib could just provide the tools to do it. It should be up to embedders to decide whether to build the async context themselves.

**Comment by gterzian** [comment]
> but I think async will make it worse

@wusyong Thanks for your input! Let's see how we can find ways to actually improve the situation in the future...

> This is easily expressed with await in async Rust

Yes I understand what you mean, and this goes straight to the core of my earlier argument: async Rust can use `await` to simplify what otherwise would require some kind of state machine, but this is because the async Rust is "the code run by the runtime". But in our case our Rust code **is the runtime**, so instead of calling `await`, we have to keep the HTML loop running and track state changes with a state machine. It is the JS we run that can remain unaware of that.

**Comment by gterzian** [comment]
Following the changes at https://github.com/servo/servo/pull/34486, we should:

- [ ] Have the compositor drive all updates of the rendering, not just those for animations.
- [ ] Remove all other use of [`note_rendering_opportunity`](https://github.com/servo/servo/blob/de8a8062bb7324bc9cb65614f2f8f6f9ed1de575/components/script/script_thread.rs#L811), since a compositor-driven workflow should make them redundant.

**Comment by gterzian** [comment]
Came across this old issue which seems relevant to updating the rendering and how we do animations, but I am not sure how much it is still relevant today, to be investigated: https://github.com/servo/servo/issues/25305. cc @mrobinson

**Comment by gterzian** [comment]
FYI there is a new related issue at https://github.com/servo/servo/issues/34637

**Comment by KiChjang** [comment]
Okay so with the help of generative AI, I now understand what was being discussed.

First of all, I don't actually mean to import an existing async runtime, and I don't think it's necessary either. I see an opportunity for us to simply reuse the script thread to act as a runtime -- one that can poll and drive `Future`s into completion. We don't even need concurrency or asynchronicity at all: we'd just simply be leveraging async Rust syntax to generate state machines for us, and have the script thread poll them into completion. Think of it as adapting async Rust syntax to express HTML event loops and tasks.

Does this make sense?
